ccmbioinfo/lfs
# CARDS-2055: The generate_compose_yaml.py script should support using a singular Percona Server for MongoDB Docker container for data persistence

This PR implements CARDS-2055 by adding support to `generate_compose_yaml.py` for generating Docker Compose environments where data persistence is handled by a _Percona Server for MongoDB_ Docker container. Options for configuring Percona encryption-at-rest are available where the encryption-at-rest key can be retrieved from a HashiCorp Vault server.

### Testing Instructions

- Before performing the tests described below, build this (`CARDS-2055`) branch, including a Docker image with `mvn clean install -Pdocker`.

- Each of these tests should be executed in the `compose-cluster` directory.

#### Single, Unencrypted Percona (with data stored in a Docker volume)

1. `python3 generate_compose_yaml.py --dev_docker_image --percona_singular`
2. `docker-compose build`
3. `docker-compose up -d`
4. Visit http://localhost:8080, login as `admin`:`admin`, create a test _Patient_ Subject.
5. `docker-compose down && docker-compose rm`
6. `docker-compose up -d`
7. Visit http://localhost:8080, login as `admin`:`admin`, the test _Patient_ Subject should still be present.
8. `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`

#### Single, Unencrypted Percona (with data stored in a file system mount)

1. `mkdir ~/tmp_percona_db`
2. `sudo chown 1001 ~/tmp_percona_db`
3. `python3 generate_compose_yaml.py --dev_docker_image --percona_singular --data_db_mount ~/tmp_percona_db`
4. `docker-compose build`
5. `docker-compose up -d`
6. Visit http://localhost:8080, login as `admin`:`admin`, create a test _Patient_ Subject.
7. `docker-compose down && docker-compose rm`
8. `docker-compose up -d`
9. Visit http://localhost:8080, login as `admin`:`admin`, the test _Patient_ Subject should still be present.
10. `docker run --rm -u 1001 -v $(realpath ~/tmp_percona_db):/data:ro -it alpine:3.14`
11. `cd /data`
12. Run `strings * | grep cards`. This should return data.
13. Run `strings * | grep 'encrypted=false'`. This should return data.
14. Run `strings * | grep 'encrypted=true'`. This should _not_ return any data.
15. `exit`
16. `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
17. `sudo rm -rf ~/tmp_percona_db`

#### Single, Encrypted (using key file) Percona (with data stored in a Docker volume)

1. `openssl rand -base64 32 > ~/tmp_percona_keyfile.key`
2. `sudo chown 1001 ~/tmp_percona_keyfile.key`
3. `sudo chmod 600 ~/tmp_percona_keyfile.key`
4. `python3 generate_compose_yaml.py --dev_docker_image --percona_singular --percona_encryption_keyfile ~/tmp_percona_keyfile.key`
5. `docker-compose build`
6. `docker-compose up -d`
7. Visit http://localhost:8080, login as `admin`:`admin`, create a test _Patient_ Subject.
8. `docker-compose down && docker-compose rm`
9. `docker-compose up -d`
10. Visit http://localhost:8080, login as `admin`:`admin`, the test _Patient_ Subject should still be present.
11. `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
12. `sudo rm ~/tmp_percona_keyfile.key`

#### Single, Encrypted (using key file) Percona (with data stored in a file system mount)

1. `openssl rand -base64 32 > ~/tmp_percona_keyfile.key`
2. `sudo chown 1001 ~/tmp_percona_keyfile.key`
3. `sudo chmod 600 ~/tmp_percona_keyfile.key`
4. `mkdir ~/tmp_percona_db`
5. `sudo chown 1001 ~/tmp_percona_db`
6. `python3 generate_compose_yaml.py --dev_docker_image --percona_singular --percona_encryption_keyfile ~/tmp_percona_keyfile.key --data_db_mount ~/tmp_percona_db`
7. `docker-compose build`
8. `docker-compose up -d`
9. Visit http://localhost:8080, login as `admin`:`admin`, create a test _Patient_ Subject.
10. `docker-compose down && docker-compose rm`
11. `docker-compose up -d`
12. Visit http://localhost:8080, login as `admin`:`admin`, the test _Patient_ Subject should still be present.
13. `docker run --rm -u 1001 -v $(realpath ~/tmp_percona_db):/data:ro -it alpine:3.14`
14. `cd /data`
15. Run `strings * | grep cards`. This should _not_ return any data.
16. Run `strings * | grep 'encrypted=false'`. This should _not_ return any data.
17. Run `strings * | grep 'encrypted=true'`. This should return data.
18. `exit`
19. `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
20. `sudo rm -rf ~/tmp_percona_db`
21. `sudo rm -rf ~/tmp_percona_keyfile.key`

#### Single, Encrypted (using Vault provided key) Percona (with data stored in a Docker volume)

1. `sudo chown 1001 vault_dev_token.token`
2. `sudo chmod 600 vault_dev_token.token`
3. `./quick-environments/percona-vault-ci.sh`
4. Visit http://localhost:8200, login with a _Token_ of `vault_dev_token`, go to _secret_, and check that the `perconaencrypt` secret is present.
5. Visit http://localhost:8080, login as `admin`:`admin`, create a test _Patient_ Subject.
6. `docker-compose down && docker-compose rm`
7. `./quick-environments/percona-vault-ci.sh`
8. Visit http://localhost:8080, login as `admin`:`admin`, the test _Patient_ Subject should still be present.
9. `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`

#### Single, Encrypted (using Vault provided key) Percona (with data stored in a file system mount)

1. `mkdir ~/tmp_percona_db`
2. `sudo chown 1001 ~/tmp_percona_db`
3. `sudo chown 1001 vault_dev_token.token`
4. `sudo chmod 600 vault_dev_token.token`
5. `./quick-environments/percona-vault-ci.sh --data_db_mount ~/tmp_percona_db`
6. Visit http://localhost:8200, login with a _Token_ of `vault_dev_token`, go to _secret_, and check that the `perconaencrypt` secret is present.
7. Visit http://localhost:8080, login as `admin`:`admin`, create a test _Patient_ Subject.
8. `docker-compose down && docker-compose rm`
9. `./quick-environments/percona-vault-ci.sh --data_db_mount ~/tmp_percona_db`
10. Visit http://localhost:8080, login as `admin`:`admin`, the test _Patient_ Subject should still be present.
11. `docker run --rm -u 1001 -v $(realpath ~/tmp_percona_db):/data:ro -it alpine:3.14`
12. `cd /data`
13. Run `strings * | grep cards`. This should _not_ return any data.
14. Run `strings * | grep 'encrypted=false'`. This should _not_ return any data.
15. Run `strings * | grep 'encrypted=true'`. This should return data.
16. `exit`
17. `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
18. `sudo rm -rf ~/tmp_percona_db`
